### PR TITLE
update 2.0 public api in nn

### PR DESCRIPTION
**Pull request by zhiboniu**
### PR types

Others
### PR changes

APIs
### Describe

update 2.0 public api in nn

背景：
对于用户调用api的层级，由于之前没有明确的设计，api开放层级比较随意。影响用户使用中的逻辑和正规感受。
目的：
1）重新整理api开放层级，提升用户使用感受
2）API保留自有特色，同时更好地符合行业规范，降低用户学习成本
3）提高框架api逻辑性，便于后续维护和优化
主要实现方法：
1）__all__内容清晰化，内容逐个列出，一行一个。不再使用 __all__ 的向上传递机制，仅在公开API层级添加 __all__，其余的 __all__ 列表统一置空
2）不再使用 #DEFINE_ALIAS
3）导入内容清晰化，原则上不再使用 import *， 按需引入需要的API， 每行一个
4）原则上不再使用 from . import module， from .module import submodule时会引入module，避免重复引用
5）添加新的Public API时，需向上寻找到最近的节点添加 import 并加入 __all__ 列表
6）Public API列表由多个公开API层级的 __all__ 列表共同组成，文档抽取以此为准，这些指定层级是：
- paddle
- paddle.amp
- paddle.nn
- paddle.nn.functional
- paddle.nn.initializer
- paddle.nn.utils
- paddle.static
- paddle.static.nn
- paddle.io
- paddle.jit
- paddle.metric
- paddle.distribution
- paddle.optimizer
- paddle.optimizer.lr
- paddle.regularizer
- paddle.text
- paddle.vision
- paddle.callbacks
- paddle.distributed
- paddle.distributed.fleet
- paddle.distributed.fleet.base.distributed_strategy
- paddle.distributed.fleet.data_generator
- paddle.distributed.fleet.utils
- paddle.distributed.parallel
- paddle.distributed.utils
- paddle.sysconfig
- paddle.utils
- paddle.utils.download
- paddle.utils.profiler
修改示例展示：
原来写法：
image
新的写法：
image

对开发者使用API的影响：
在指定层级推荐方式：（不受影响）
  from paddle.nn import CTCLoss
  from paddle.nn import Conv2D
非指定层级支持但不推荐：（不受影响）
  from paddle.nn.layer.loss import CTCLoss
  from paddle.nn.layer.conv import Conv2D
对于import *使用方式不推荐。（非公开层级受影响）
  不再支持：（非公开层级）
  from paddle.nn.layer import *
  from paddle.nn.layer.conv import *
  from paddle.nn.initializer.norm import *
  from paddle.tensor.math import *
  from paddle.optimizer.adam import *
  可以支持：（公开层级）
  from paddle import *
  from paddle.nn import *
  from paddle.utils import *
  from paddle.optimizer import *

本pr涉及对api在__all___中暴露的相关改动：
paddle.nn 层级改动：
1）'diag_embed'  根据设计文档仅在paddle.nn.funcitional中公开，此处__all__中移除。
2）'remove_weight_norm', 'weight_norm' remove to utils 根据设计文档在paddle.nn.utils中公开，此处__all__中移除。
3）新增以下nn功能，在该层级公开
{'Softmax', 'Hardtanh', 'LogSoftmax', 'clip_by_norm', 'ReLU6', 'Swish', 'Sequential', 'AdaptiveMaxPool2D', 
'TransformerEncoderLayer', 'ReLU', 'MSELoss', 'SELU', 'Hardswish', 'Hardsigmoid', 'Conv3DTranspose', 
'MaxPool3D', 'AdaptiveAvgPool1D', 'ELU', 'SmoothL1Loss', 'Conv2DTranspose', 'dynamic_decode', 'Conv2D', 
'LeakyReLU', 'HSigmoidLoss', 'Sigmoid', 'Hardshrink', 'Conv1DTranspose', 'AdaptiveMaxPool3D', 'ClipGradByValue', 
'Flatten', 'L1Loss', 'LayerList', 'AvgPool1D', 'Tanhshrink', 'PairwiseDistance', 'Conv1D', 'AdaptiveMaxPool1D', 
'MultiHeadAttention', 'CTCLoss', 'TransformerDecoder', 'CrossEntropyLoss', 'LogSigmoid', 'TransformerEncoder', 
'ParameterList', 'GELU', 'MarginRankingLoss', 'MaxPool1D', 'TransformerDecoderLayer', 'Softsign', 'Layer', 'clip', 
'PixelShuffle', 'AdaptiveAvgPool2D', 'Softplus', 'NLLLoss', 'BCEWithLogitsLoss', 'KLDivLoss', 'UnfoldRNNCellBase', 
'ThresholdedReLU', 'Conv3D', 'Softshrink', 'AvgPool3D', 'PReLU', 'Maxout', 'AdaptiveAvgPool3D', 'MaxPool2D', 
'BeamSearchDecoder', 'AvgPool2D', 'ClipGradByGlobalNorm', 'BCELoss', 'ClipGradByNorm', 'Transformer', 'Tanh'}

示例一，网络相关api默认在paddle.nn开放。以调用CTCLoss、Conv2D为例：
  升级前：
  可以使用方式：
  from paddle.nn.layer.loss import CTCLoss
  from paddle.nn.layer.conv import Conv2D
  from paddle.nn.layer.loss import *
  from paddle.nn.layer.conv import *
  from paddle.nn import CTCLoss
  from paddle.nn import Conv2D
  不支持使用方式：
  from paddle.nn import *
  升级后：
  默认推荐使用：
  from paddle.nn import CTCLoss
  from paddle.nn import Conv2D
  可以运行但不推荐使用：
  from paddle.nn import *
  from paddle.nn.layer.loss import CTCLoss
  from paddle.nn.layer.conv import Conv2D
  将不再支持：
  from paddle.nn.layer.loss import *
  from paddle.nn.layer.conv import *

示例二，原部分api在多处开放现默认在paddle.nn开放。以调用BatchNorm2D为例：
  升级前：
  可以使用方式：
  from paddle.nn.layer.norm import BatchNorm2D
  from paddle.nn.layer.norm import *
  from paddle.nn import BatchNorm2D
  from paddle.nn import *
  升级后：
  默认推荐方式：
  from paddle.nn import BatchNorm2D
  可以使用但不推荐：
  from paddle.nn.layer.norm import BatchNorm2D
  from paddle.nn import *
  将不再支持：
  from paddle.nn.layer.norm import *

示例三，个别api原在多处开放，现默认在专用模块开放。对于'remove_weight_norm', 'weight_norm', 'diag_embed'三个api
  升级前：
  可以使用方式：
  from paddle.nn.functional import diag_embed
  from paddle.nn.functional import *
  from paddle.nn import diag_embed
  from paddle.nn import *
  升级后：
  推荐使用方式：
  from paddle.nn.functional import diag_embed
  from paddle.nn.utils import remove_weight_norm
  from paddle.nn.utils import weight_norm
  可以使用但不推荐：
  from paddle.nn import diag_embed
  from paddle.nn import remove_weight_norm
  from paddle.nn import weight_norm
  不再支持方式：
  from paddle.nn import *

### CODE REVIEW

File: python/paddle/nn/__init__.py [hunk 1]
```diff
@@ -15,144 +15,232 @@
 # TODO: import all neural network related api under this directory,
 # including layers, linear, conv, rnn etc.
 
-from .layer import norm
-from .functional import extension
-from .layer import common
-from .layer import rnn
-from .utils import weight_norm_hook
+from .layer import norm  # noqa: F401
```
Comment (willthefrog) [inline python/paddle/nn/__init__.py]: double check these 5

File: python/paddle/nn/functional/common.py [hunk 2]
```diff
@@ -20,44 +20,22 @@
...
-from ...fluid.layers import unsqueeze  #DEFINE_ALIAS
+from ...fluid.layers import unfold  # noqa: F401
+from ...fluid.layers import squeeze  # noqa: F401
+from ...fluid.layers import unsqueeze  # noqa: F401
 from ...tensor import clip
 from ...tensor import sum
 from ...tensor import sqrt
-from ...tensor import sum  #DEFINE_ALIAS
-from ...tensor import sqrt  #DEFINE_ALIAS
+from ...tensor import sum  # noqa: F401
```
Comment (willthefrog) [inline python/paddle/nn/functional/common.py]: duplicated, also, these imports are actually used in the python file, so no need for `noqa`

File: python/paddle/nn/functional/loss.py [hunk 3]
```diff
@@ -23,41 +23,20 @@
...
-from ...fluid.layers import dice_loss  #DEFINE_ALIAS
-from ...fluid.layers import log_loss  #DEFINE_ALIAS
-from ...fluid.layers import npair_loss  #DEFINE_ALIAS
+from ...fluid.layers import dice_loss  # noqa: F401
+from ...fluid.layers import log_loss  # noqa: F401
+from ...fluid.layers import npair_loss  # noqa: F401
 from ...fluid.layers import reshape
-from ...fluid.layers import softmax_with_cross_entropy  #DEFINE_ALIAS
-from ...fluid.layers import square_error_cost  #DEFINE_ALIAS
+from ...fluid.layers import softmax_with_cross_entropy  # noqa: F401
```
Comment (willthefrog) [inline python/paddle/nn/functional/loss.py]: remove `noqa`, `softmax_with_cross_entropy` is actually used, double check others

File: python/paddle/nn/layer/norm.py [hunk 4]
```diff
@@ -28,13 +28,10 @@
 # TODO: define normalization api  
 
 import six
-#from ...fluid.dygraph.nn import InstanceNorm
 
-from ...fluid.dygraph import BatchNorm  #DEFINE_ALIAS
-#from ...fluid.dygraph import GroupNorm  #DEFINE_ALIAS
+from ...fluid.dygraph import BatchNorm  # noqa: F401
```
Comment (willthefrog) [inline python/paddle/nn/layer/norm.py:32]: `BatchNorm` is no longer needed, export `BatchNorm1d`- 3d instead

**Review by willthefrog — CHANGES_REQUESTED** [review]

File: python/paddle/nn/__init__.py (hunk 1 above)
Comment (zhiboniu) [reply python/paddle/nn/__init__.py]: done

File: python/paddle/nn/functional/common.py (hunk 2 above)
Comment (zhiboniu) [reply python/paddle/nn/functional/common.py]: done

File: python/paddle/nn/functional/loss.py (hunk 3 above)
Comment (zhiboniu) [reply python/paddle/nn/functional/loss.py]: done

File: python/paddle/nn/layer/norm.py (hunk 4 above)
Comment (zhiboniu) [reply python/paddle/nn/layer/norm.py:32]: done

Comment (zhiboniu) [reply python/paddle/nn/layer/norm.py:32]: export BatchNorm1d- 3d, but keep BatchNorm also to compatible with old usage

File: python/paddle/nn/functional/__init__.py [hunk 5]
```diff
@@ -14,208 +14,174 @@
...
-from .pooling import max_pool3d  #DEFINE_ALIAS
+from .common import pad  # noqa: F401
+from .common import cosine_similarity  # noqa: F401
+from .common import unfold  # noqa: F401
+from .common import interpolate  # noqa: F401
+from .common import upsample  # noqa: F401
+from .common import bilinear  # noqa: F401
+from .conv import conv1d  # noqa: F401
+from .conv import conv1d_transpose  # noqa: F401
+from .common import linear  # noqa: F401
```
Comment (willthefrog) [inline python/paddle/nn/functional/__init__.py:60]: move up

**Review by willthefrog — CHANGES_REQUESTED** [review]

File: python/paddle/nn/functional/__init__.py (hunk 5 above)
Comment (zhiboniu) [reply python/paddle/nn/functional/__init__.py:60]: done

File: python/paddle/nn/__init__.py [hunk 6]
```diff
@@ -15,144 +15,227 @@
...
+from .layer.vision import PixelShuffle  # noqa: F401
 
-from .layer.transformer import MultiHeadAttention
-from .layer.transformer import TransformerEncoderLayer
-from .layer.transformer import TransformerEncoder
-from .layer.transformer import TransformerDecoderLayer
-from .layer.transformer import TransformerDecoder
-from .layer.transformer import Transformer
-from .layer.distance import PairwiseDistance  #DEFINE_ALIAS
+from .layer import loss  # noqa: F401
```
Comment (willthefrog) [inline python/paddle/nn/__init__.py:130]: 为什么要引入这几个子模块？

File: python/paddle/nn/initializer/__init__.py [hunk 7]
```diff
@@ -13,36 +13,33 @@
 # limitations under the License.
 
 # TODO: define the initializers to create a Parameter in neural network
-from ...fluid.initializer import Bilinear  #DEFINE_ALIAS
-from ...fluid.initializer import set_global_initializer  #DEFINE_ALIAS
+from ...fluid.initializer import Bilinear  # noqa: F401
+from ...fluid.initializer import set_global_initializer  # noqa: F401
```
Comment (willthefrog) [inline python/paddle/nn/initializer/__init__.py:17]: 这个有用到吗？

**Review by willthefrog — CHANGES_REQUESTED** [review]

File: python/paddle/nn/__init__.py (hunk 6 above)
Comment (zhiboniu) [reply python/paddle/nn/__init__.py:127]: 漏删了，已删

File: python/paddle/nn/initializer/__init__.py (hunk 7 above)
Comment (zhiboniu) [reply python/paddle/nn/initializer/__init__.py:17]: 考虑到专门从fluid引过来的，而且没有相同功能的api，先保留了。已删

File: python/paddle/nn/__init__.py (hunk 6 above)
Comment (zhiboniu) [reply python/paddle/nn/__init__.py:130]: 其他两个子模块删了，nn.loss在unitest里用的非常多，暂时先保留了

**Review by willthefrog — DISMISSED** [review]

File: python/paddle/nn/__init__.py [hunk 8]
```diff
@@ -15,147 +15,232 @@
...
+           'BeamSearchDecoder',
+           'ClipGradByNorm',
+           'ReLU',
+           'PairwiseDistance',
+           'BCEWithLogitsLoss',
+           'SmoothL1Loss',
+           'MaxPool3D',
+           'AdaptiveMaxPool2D',
+           'Hardshrink',
+           'clip',
```
Comment (iclementine) [inline python/paddle/nn/__init__.py:230]: 子模块不需要写在 __all__ 这里吗？
有 clip, 但是没有 layer 和 functional?

**Review by iclementine — DISMISSED** [review]
LGTM

File: python/paddle/nn/__init__.py (hunk 8 above)
Comment (zhiboniu) [reply python/paddle/nn/__init__.py:230]: 子模块layer里的api已经包含了，就不需要再包含子模块了吧，感觉重复了。functional里有自己单独的all。我觉得其实这个clip应该也可以不包含吧，已经包含了它里面的api了。__all__的内容应该保留至api级别吧

**Review by jzhang533 — DISMISSED** [review]
LGTM

**Review by XiaoguangHu01 — DISMISSED** [review]
LGTM

**Review by willthefrog — DISMISSED** [review]
LGTM

**Review by raindrops2sea — APPROVED** [review]

**Review by jzhang533 — APPROVED** [review]
lgtm

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM
